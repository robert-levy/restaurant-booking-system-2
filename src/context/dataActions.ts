import { State, IBookingRequest } from "../interfaces/interfaces";

export const initialState: State = {
  bar: [],
  tables: [],
};

export enum ActionTypes {
  CREATE_RESTAURANT = "CREATE_RESTAURANT",
  MAKE_BOOKING = "MAKE_BOOKING",
  MAKE_TABLE_AVAILABLE = "MAKE_TABLE_AVAILABLE",
  ACKNOWLEDGE_ERROR = "ACKNOWLEDGE_ERROR",
}

interface CreateRestaurantAction {
  type: ActionTypes.CREATE_RESTAURANT;
  payload: State;
}

interface MakeBookingAction {
  type: ActionTypes.MAKE_BOOKING;
  payload: IBookingRequest;
}

interface MakeTableAvailableAction {
  type: ActionTypes.MAKE_TABLE_AVAILABLE;
  payload: number;
}

interface AcknowledgeErrorAction {
  type: ActionTypes.ACKNOWLEDGE_ERROR;
}

export type Action =
  | CreateRestaurantAction
  | MakeBookingAction
  | MakeTableAvailableAction
  | AcknowledgeErrorAction;

/// Actions
export const createRestaurant = (initState: State): CreateRestaurantAction => ({
  type: ActionTypes.CREATE_RESTAURANT,
  payload: initState,
});

export const makeBooking = (
  totalPersons: number,
  personsWithDisability: number
): MakeBookingAction => ({
  type: ActionTypes.MAKE_BOOKING,
  payload: {
    totalPersons,
    personsWithDisability,
  },
});

export const makeTableAvailable = (
  tableNumber: number
): MakeTableAvailableAction => ({
  type: ActionTypes.MAKE_TABLE_AVAILABLE,
  payload: tableNumber,
});

export const acknowledgeError = (): AcknowledgeErrorAction => ({
  type: ActionTypes.ACKNOWLEDGE_ERROR
})
