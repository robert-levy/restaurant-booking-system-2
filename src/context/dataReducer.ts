import Restaurant from "../Restaurant";
import { State } from "../interfaces/interfaces";
import { Action, ActionTypes } from "./dataActions";

const reducer = (state: State, action: Action): State => {
  let restaurant = new Restaurant();
  // let updatedState: Partial<State> = {}; or ErrorResponse
  let updatedState: any;

  switch (action.type) {
    case ActionTypes.CREATE_RESTAURANT:
      return {
        ...state,
        ...action.payload,
      };
    case ActionTypes.MAKE_BOOKING:
      updatedState = restaurant.makeBooking(state, action.payload);
      console.log(updatedState);
      if ("errorMessage" in updatedState) {
        console.log("inside");
        const { errorMessage } = updatedState;
        return {
          ...state,
          errorMessage,
        };
      }

      return { ...updatedState };

    case ActionTypes.MAKE_TABLE_AVAILABLE:
      updatedState = restaurant.makeTableAvailable(
        state.tables,
        action.payload
      );
      return {
        ...state,
        tables: updatedState, // as IRestaurantTable[], // Type assertion to State['tables']
      };

    case ActionTypes.ACKNOWLEDGE_ERROR:
      updatedState = { ...state };
      delete updatedState.errorMessage;
      return {
        ...updatedState,
      };
    default:
      return state;
  }
};
export default reducer;
