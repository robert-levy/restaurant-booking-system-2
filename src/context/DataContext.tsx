import React, { createContext, useReducer, useContext } from "react";
import dataReducer from "./dataReducer";
import { initialState, Action } from "./dataActions";
import { State } from "../interfaces/interfaces";

export interface DataContextValue {
  state: State;
  dispatch: React.Dispatch<Action>;
}

const DataContext = createContext<DataContextValue | undefined>(undefined);

interface DataProviderProps {
  children: React.ReactNode;
}

export const useDataDispatch = () => {
  const context = useContext(DataContext);
  if (context === undefined) {
    throw new Error(`useDispatch must be used within a Provier`);
  }
  return context;
};

export const DataProvider: React.FC<DataProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(dataReducer, initialState);

  return (
    <DataContext.Provider value={{ state, dispatch }}>
      {children}
    </DataContext.Provider>
  );
};
