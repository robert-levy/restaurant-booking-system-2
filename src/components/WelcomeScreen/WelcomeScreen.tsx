import React from "react";
import { Paper, Typography, TextField, IconButton } from "@mui/material/";
import {useDataDispatch} from "../../context/DataContext";
import RestaurantTablesJson from '../../RestaurantTables.json'
import RestaurantBarJson from '../../RestaurantBar.json'
import { createRestaurant } from "../../context/dataActions";
import CheckCircleIcon from "@mui/icons-material/CheckCircle";
import styled from "@emotion/styled";
import { IBarSeat, IRestaurantTable } from "../../interfaces/interfaces";

interface IWelcomeModal {
  setBranchName: React.Dispatch<React.SetStateAction<string>>;
}

const StyledWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

const StyledPaper = styled(Paper)`
  background: #e5e5e5;
  position: absolute;
  height: 300px;
  padding: 25px;
  @media (max-width: 350px) {
    height: 500px;
  }
`;

const FormWrapper = styled.form`
  display: flex;
  justify-content: center;
  position: relative;
  top: 10%;
  gap:20px;
  @media (max-width: 350px) {
    flex-direction: column;
  }
`;

const WelcomeScreen: React.FC<IWelcomeModal> = ({
  setBranchName,
}) => {
  const {dispatch} = useDataDispatch()

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    // setup state
    dispatch(createRestaurant({
      tables: RestaurantTablesJson as IRestaurantTable[],
      bar: RestaurantBarJson as IBarSeat[]
    }))

    const inputElement = e.currentTarget.elements.namedItem(
      "restaurant-name-field"
    ) as HTMLInputElement;
    const inputValue = inputElement.value;
    setBranchName(inputValue)
  };

  return (
    <StyledWrapper>
      <StyledPaper elevation={24}>
        <Typography align="center" variant="h5">
          Welcome to Restuarant Booking System
        </Typography>
        <Typography
          variant="subtitle1"
          marginTop={3}
          marginBottom={2}
          align="center"
        >
          Please enter a restaurant name to begin
        </Typography>
        <FormWrapper onSubmit={handleSubmit}>
          <TextField
            required
            id="restaurant-name-field"
            label="Required"
            defaultValue="Nandos Shoreditch"
          />
          <IconButton aria-label="CheckCircleIcon" type="submit">
            <CheckCircleIcon color="secondary" fontSize="large" />
          </IconButton>
        </FormWrapper>
      </StyledPaper>
    </StyledWrapper>
  );
};

export default WelcomeScreen;
