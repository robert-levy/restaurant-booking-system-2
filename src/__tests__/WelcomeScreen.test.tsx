import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import WelcomeScreen from "../components/WelcomeScreen/WelcomeScreen";
import { DataProvider } from "../context/DataContext";

describe("WelcomeScreen.tsx", () => {
  const setBranchNameMock = jest.fn();
  render(
    <DataProvider>
      <WelcomeScreen setBranchName={setBranchNameMock} />
    </DataProvider>
  );

  test("snapshot test", () => {
    const component = renderer.create(
      <DataProvider>
        <WelcomeScreen setBranchName={setBranchNameMock} />
      </DataProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
