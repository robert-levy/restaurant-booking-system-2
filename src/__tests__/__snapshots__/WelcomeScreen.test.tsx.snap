// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomeScreen.tsx snapshot test 1`] = `
<div
  className="css-1q2lbmy"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 css-1oyklf4-MuiPaper-root"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
    >
      Welcome to Restuarant Booking System
    </h5>
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignCenter css-16clf7q-MuiTypography-root"
    >
      Please enter a restaurant name to begin
    </h6>
    <form
      className="css-6batyf"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-testid="restaurant-name-input-test"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={true}
          htmlFor="restaurant-name-field"
          id="restaurant-name-field-label"
        >
          Required
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            defaultValue="Nandos Shoreditch"
            disabled={false}
            id="restaurant-name-field"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-14lo706"
            >
              <span>
                Required
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        aria-label="CheckCircleIcon"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="welcome-screen-submit"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge css-1nryff-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
